core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -246,14 +255,78 @@ class UserService(
+    if (billingProject.status != CreationStatuses.CreatingLandingZone) return billingProject
+    val record = Await.result(
+      workspaceResourceRecordDao.selectByBillingProject(billingProject.projectName).map(_.headOption),
+      DurationInt(30).seconds
+    )
+
+    def updateLandingZoneSuccess(lzId: UUID): RawlsBillingProject = {
+      billingRepository.updateLandingZoneId(billingProject.projectName, lzId)
+      billingRepository.updateCreationStatus(billingProject.projectName, CreationStatuses.Ready, None)
+      record.foreach(workspaceResourceRecordDao.delete)
+      billingProject.copy(status = CreationStatuses.Ready, landingZoneId = Some(lzId.toString))
+    }
+
+    def updateLandingZoneFailure(msg: String): RawlsBillingProject = {
+      val message = Some(s"Landing Zone creation failed: $msg")
+      billingRepository.updateCreationStatus(billingProject.projectName, CreationStatuses.Error, message)
+      record.foreach(workspaceResourceRecordDao.delete)
+      billingProject.copy(status = CreationStatuses.Error, message = message)
+    }
+
+    try
+      record.map(r => workspaceManagerDAO.getCreateAzureLandingZoneResult(r.jobControlId.toString, ctx)) match {
+        case None => updateLandingZoneFailure("No monitoring record available")
+        // we have a status - this should be easy
+        case Some(result) if result.getJobReport != null && result.getJobReport.getStatus != null =>
+          result.getJobReport.getStatus match {
+            // the job just isn't done yet - return the billing project unchanged
+            case JobReport.StatusEnum.RUNNING => billingProject
+            case JobReport.StatusEnum.FAILED =>
+              updateLandingZoneFailure(
+                Option(result.getErrorReport).map(_.getMessage).getOrElse("Failure Reported, but no errors returned")
+              )
+            case JobReport.StatusEnum.SUCCEEDED if result.getLandingZone == null =>
+              updateLandingZoneFailure("Landing Zone result marked as successful, but no landing zone returned")
+            case JobReport.StatusEnum.SUCCEEDED => updateLandingZoneSuccess(result.getLandingZone.getId)
+          }
+        case Some(result) if result.getLandingZone != null && result.getLandingZone.getId != null =>
+          updateLandingZoneSuccess(result.getLandingZone.getId)
+        case Some(result) if result.getErrorReport != null =>
+          updateLandingZoneFailure(result.getErrorReport.getMessage)
+        // no job report, no landing zone, and no error report
+        // return project with status as error, but don't update anything in the database, so it retries in the future, and no data is lost for now
+        case Some(_) =>
+          billingProject.copy(status = CreationStatuses.Error,
+                              message = Some("Unable to retrieve landing zone results")
+          )
+      }
+    catch {
+      case e: ApiException =>
```

Comment:
Should we allow 5xx errors to be retry-able and thus not push the billing project status to error? That is, if WSM is temporarily unavailable it would be good to be resilient to that situation and allow the billing project to retry.

Reply:
It doesn't persist the error state, so it will re-attempt in future requests.
I figured since this was a user request that UIs could be waiting on, it might be make sense to avoid spending time in retries. It probably does make sense for 5XX though.

Reply:
I suggest that even for 5XXs, this is not worth it here:

- The existing trait that provides retry functionality elsewhere requires an `ActorSystem`
- Even if it would work, shoe-horning it in here makes for really convoluted syntax (I just spent way too much time trying to make it work - it was kind of hideous)
- Re-implementing it with decent testing, correct config of timeouts and n-retries, etc would take more time than we probably want to spend

After writting the above, I found the Retry that doesn't require the ActorSystem (it's in a companion object at the bottom of the file). I have a reasonable solution that uses it, but it has two drawbacks:
1. it extends the unit tests by however long you set your timeout
2. there's no way of only retrying some exceptions - it's all or nothing